src/libraries/System.Net.Security/src/System/Net/Security/SslConnectionInfo.Unix.cs

```diff
@@ -12,14 +17,2710 @@ internal partial struct SslConnectionInfo
+                    DataKeySize = 256;
+                    DataHashAlg = (int)HashAlgorithmType.None;
+                    DataHashKeySize = 0;
+                    break;
+
+                case TlsCipherSuite.TLS_ECDHE_PSK_WITH_CHACHA20_POLY1305_SHA256:
+                    KeyExchangeAlg = (int)ExchangeAlgorithmType.DiffieHellman;
+                    DataCipherAlg = (int)CipherAlgorithmType.None;
+                    DataKeySize = 256;
+                    DataHashAlg = (int)HashAlgorithmType.None;
+                    DataHashKeySize = 0;
+                    break;
+
+                case TlsCipherSuite.TLS_DHE_PSK_WITH_CHACHA20_POLY1305_SHA256:
+                    KeyExchangeAlg = (int)ExchangeAlgorithmType.DiffieHellman;
+                    DataCipherAlg = (int)CipherAlgorithmType.None;
+                    DataKeySize = 256;
+                    DataHashAlg = (int)HashAlgorithmType.None;
+                    DataHashKeySize = 0;
+                    break;
+
+                case TlsCipherSuite.TLS_RSA_PSK_WITH_CHACHA20_POLY1305_SHA256:
+                    KeyExchangeAlg = (int)ExchangeAlgorithmType.RsaKeyX;
+                    DataCipherAlg = (int)CipherAlgorithmType.None;
+                    DataKeySize = 256;
+                    DataHashAlg = (int)HashAlgorithmType.None;
+                    DataHashKeySize = 0;
+                    break;
+
+                case TlsCipherSuite.TLS_ECDHE_PSK_WITH_AES_128_GCM_SHA256:
+                    KeyExchangeAlg = (int)ExchangeAlgorithmType.DiffieHellman;
+                    DataCipherAlg = (int)CipherAlgorithmType.Aes128;
+                    DataKeySize = 128;
+                    DataHashAlg = (int)HashAlgorithmType.None;
+                    DataHashKeySize = 0;
+                    break;
+
+                case TlsCipherSuite.TLS_ECDHE_PSK_WITH_AES_256_GCM_SHA384:
+                    KeyExchangeAlg = (int)ExchangeAlgorithmType.DiffieHellman;
+                    DataCipherAlg = (int)CipherAlgorithmType.Aes256;
+                    DataKeySize = 256;
+                    DataHashAlg = (int)HashAlgorithmType.None;
+                    DataHashKeySize = 0;
+                    break;
+
+                case TlsCipherSuite.TLS_ECDHE_PSK_WITH_AES_128_CCM_8_SHA256:
+                    KeyExchangeAlg = (int)ExchangeAlgorithmType.DiffieHellman;
+                    DataCipherAlg = (int)CipherAlgorithmType.Aes128;
+                    DataKeySize = 128;
+                    DataHashAlg = (int)HashAlgorithmType.None;
+                    DataHashKeySize = 0;
+                    break;
+
+                case TlsCipherSuite.TLS_ECDHE_PSK_WITH_AES_128_CCM_SHA256:
+                    KeyExchangeAlg = (int)ExchangeAlgorithmType.DiffieHellman;
+                    DataCipherAlg = (int)CipherAlgorithmType.Aes128;
+                    DataKeySize = 128;
+                    DataHashAlg = (int)HashAlgorithmType.None;
+                    DataHashKeySize = 0;
```

Comment:
`DataHashKeySize` was previously computed based on `DataHashAlg`. If IL size is a concern, would it help to call the helper after the switch instead of setting it from every case?

Given this PR is about improving size, have you looked at what it would mean to encode the data into a binary blob instead?
E.g. generate something along the lines of
```c#
private void MapCipherSuite(TlsCipherSuite cipherSuite)
{
    int index = Array.BinarySearch(s_cipherSuites, cipherSuite);
    Debug.Assert(index >= 0);

    index *= 4;    
    KeyExchangeAlg = TlsCipherSuiteData[offset];
    DataCipherAlg = TlsCipherSuiteData[offset + 1];
    DataKeySize = TlsCipherSuiteData[offset + 2];
    DataHashAlg = TlsCipherSuiteData[offset + 3];
    DataHashKeySize = SomeHelper(DataHashAlg);
}

// Or just a switch from TlsCipherSuite => index
private static readonly TlsCipherSuite[] s_cipherSuites = Enum.GetValues<TlsCipherSuite>();

private static ReadOnlySpan<byte> TlsCipherSuiteData => new byte[]
{
    42, 42, 42, 42 // Source generated
}
```

Reply:
> `DataHashKeySize` was previously computed based on `DataHashAlg`

It still is computed based on `DataHashAlg`, see the change in `TlsCipherSuiteNameParser.ttinclude`, we have moved its generation from run-time to build-time. This one is computed as `0`.

Reply:
Right, but what I mean is that before it was 1 call to a helper in compiled code, now it's hundreds of field assignments.

Reply:
> this is actually not about IL size

This PR should not regress IL size significantly and some data could be encoded more optimally. For example, it looks like that `DataHashKeySize` could be easily computed from `DataHashAlg` or could each case initialize only non-default values? 

e.g. `TLS_ECCPWD_WITH_AES_128_CCM_SHA256` would look like

```c#
        case TlsCipherSuite.TLS_ECCPWD_WITH_AES_128_CCM_SHA256:
            DataCipherAlg = (int)CipherAlgorithmType.Aes128;
            DataKeySize = 128;
            break;
```

Reply:
> This PR should not regress IL size significantly

In the current state, it is not regressing the IL size. It is producing ~3 KB smaller assembly compared to main and 18.3 KB smaller NativeAOT published app.

We can definitely improve it further. I'll try the your middle-ground solution and collect the stats.

Reply:
By hoisting most common defaults out like this: https://github.com/am11/runtime/commit/f50c11fda80b4044ae9aa9b3d7b7ee148e23f8fd, it is producing ~7 KB smaller assembly compared to `main` branch and 16.5 KB smaller NativeAOT published app. It is 4 KB win in IL size and 2 KB regression in NativeAOT size compared to the current state of this PR.

(Personally, this trade-off is OK since smaller IL sizes help other app models; e.g. wasm apps)

Reply:
Why not to do other defaults too? e.g. there is still over 100's of `DataCipherAlg = (int)CipherAlgorithmType.None;`

Reply:
Just did :)
https://github.com/dotnet/runtime/commit/1126b34d32d4582626e6b06a857ed496f7cb52cc

NativeAOT: goes from 16.5 to 16.3 K reduction (regression)
IL size: goes from 7.6 to 9.2 K reduction (improvement)

If this approach is looking solid, I can push changes to the PR branch.

Reply:
Looks reasonable to me. Did you experiment with using https://github.com/dotnet/runtime/blob/f66c469be184f39420291c456c3422136c09ebdf/src/libraries/System.Net.Security/src/System/Net/Security/TlsCipherSuiteData.cs#L49 to calculate `DataHashKeySize` once the data are set in the switch?

Reply:
It saves 10.2 K of IL size and 16.3 K of NativeAOT. This version is looking better, I'll push this one.

Reply:
There is still a lot of redundant information between the switch statements. We can push this even further by mapping the cipherSuite to one integer with all information and then distributing that information into the member fields.
```
        private void MapCipherSuite(TlsCipherSuite cipherSuite)
        {
            int data = cipherSuite switch
            {
                TlsCipherSuite.TLS_NULL_WITH_NULL_NULL => 0,
                TlsCipherSuite.TLS_RSA_WITH_NULL_MD5 => 2 /* RsaKeyX */ | (1 /* Md5 */ << 2);
...
                _ => -1
            };

            Debug.Assert(data != -1, $"No mapping found for cipherSuite {cipherSuite}");

            ReadOnlySpan<int> keyExchangeAlgMap = new int[] { (int)ExchangeAlgorithmType.None, (int)ExchangeAlgorithmType.RsaSign, (int)ExchangeAlgorithmType.RsaKeyX, (int)ExchangeAlgorithmType.DiffieHellman };
            KeyExchangeAlg = (ExchangeAlgorithmType)keyExchangeAlgMap[data & 3];

            ReadOnlySpan<int> dataHashAlgMap = new int[] { (int)HashAlgorithmType.None, ... };
            DataHashAlg = dataHashAlgMap[(data >> 2) & 7];

            ...
        }
```
This would shrink the IL size by another 5 K. Is this possible to do in T4 template?

Reply:
We tried something like this after @MihaZupan's feedback https://github.com/am11/runtime/commit/d47c5673b9efad6a871cefe47bafd4558874b60d, unfortunately it wasn't saving much on NativeAOT side (18.3 K vs 9.6 K reduction). I'll try your approach of packing / unpacking the integers. It should be possible with T4 templates. 

ps: for some reason, RyuJIT is producing slightly better code with old switch-case than the switch expressions.

Reply:
> @MihaZupan's feedback https://github.com/am11/runtime/commit/d47c5673b9efad6a871cefe47bafd4558874b60d,

BinarySearch comes with a lot of baggage (ArraySortHelper, etc.). My guess is that it is the reason why it did not help.

Reply:
Pushed b3d656b, it trims 14.8 K in IL size and 16.2 K from NativeAOT app. It encodes/decodes four values as ulong.

I haven't tried encoding indices as int using maps yet.

Reply:
Moving to private method 09cde7d2aad761c648f5b1c79f08ff285158fd27 (just like `GetHashSize`) saves 16.3 K IL size and 15.7 K NativeAOT. I think due to its large size JIT stopped helping with optimization which I noticed in `GetHashSize`:

> ps: for some reason, RyuJIT is producing slightly better code with old switch-case than the switch expressions.

I will test with RVA next, one for  enumval -> index and another for index -> encodedval, without using `Array.*`.

Reply:
Pushed the ROS version: 2d15328 (pending CI testing)

It shaves off 15.8 K from IL size and 32.5 K from NativeAOT (by far the best 'relative' results)